protocol/messenger_communities.go

```diff
@@ -1947,3 +1953,802 @@ func (m *Messenger) RequestExtractDiscordChannelsAndCategories(filesToImport []s
+			progressUpdates <- importProgress
+			cancel <- communityID
+			return
+		}
+
+		var msgs []*common.Message
+		for _, msg := range messagesToSave {
+			msgs = append(msgs, msg)
+		}
+
+		err = m.persistence.SaveMessages(msgs)
+		if err != nil {
+			m.cleanUpImport(communityID)
+			importProgress.AddTaskError(discord.ImportMessagesTask, discord.Error(err.Error()))
+			importProgress.StopTask(discord.ImportMessagesTask)
+			progressUpdates <- importProgress
+			return
+		}
+
+		if m.DiscordImportMarkedAsCancelled(communityID) {
+			importProgress.StopTask(discord.ImportMessagesTask)
+			progressUpdates <- importProgress
+			cancel <- communityID
+			return
+		}
+
+		err = m.persistence.SavePinMessages(pinMessagesToSave)
+		if err != nil {
+			m.cleanUpImport(communityID)
+			importProgress.AddTaskError(discord.ImportMessagesTask, discord.Error(err.Error()))
+			importProgress.StopTask(discord.ImportMessagesTask)
+			progressUpdates <- importProgress
+			return
+		}
+
+		importProgress.UpdateTaskProgress(discord.ImportMessagesTask, 1)
+		progressUpdates <- importProgress
+
+		totalAssetsCount := len(messageAttachmentsToDownload) + len(authorProfilesToSave)
+		var assetCounter discord.AssetCounter
+
+		var wg sync.WaitGroup
+
+		for id, author := range authorProfilesToSave {
+			wg.Add(1)
+			go func(id string, author *protobuf.DiscordMessageAuthor) {
+				defer wg.Done()
+				imagePayload, err := discord.DownloadAvatarAsset(author.AvatarUrl)
+				if err != nil {
+					errmsg := fmt.Sprintf("Couldn't download profile avatar '%s': %s", author.AvatarUrl, err.Error())
+					importProgress.AddTaskError(
+						discord.DownloadAssetsTask,
+						discord.Warning(errmsg),
+					)
+					progressUpdates <- importProgress
+					return
+				}
+
+				err = m.persistence.UpdateDiscordMessageAuthorImage(author.Id, imagePayload)
+				if err != nil {
+					importProgress.AddTaskError(discord.DownloadAssetsTask, discord.Warning(err.Error()))
+					progressUpdates <- importProgress
+					return
+				}
+
+				author.AvatarImagePayload = imagePayload
+				authorProfilesToSave[id] = author
+
+				assetCounter.Increase()
+				progress := float32(assetCounter.Value()) / float32(totalAssetsCount)
+				importProgress.UpdateTaskProgress(discord.DownloadAssetsTask, progress)
+				progressUpdates <- importProgress
+
+				if m.DiscordImportMarkedAsCancelled(discordCommunity.IDString()) {
+					importProgress.StopTask(discord.DownloadAssetsTask)
+					progressUpdates <- importProgress
+					cancel <- discordCommunity.IDString()
+					return
+				}
+			}(id, author)
+		}
+		wg.Wait()
+
+		if m.DiscordImportMarkedAsCancelled(communityID) {
+			importProgress.StopTask(discord.DownloadAssetsTask)
+			progressUpdates <- importProgress
+			cancel <- communityID
+			return
+		}
+
+		for idxRange := range gopart.Partition(len(messageAttachmentsToDownload), 20) {
+			attachments := messageAttachmentsToDownload[idxRange.Low:idxRange.High]
+			wg.Add(1)
```

Comment:
@Samyoul I've looked into worker pools (finally) and realized that it's very similar to what I'm already doing.
However, I'm happy to play around with this in a second iteration and do some profiling to see whether this improves performance.

Also notice here that I've hardcoded the partitions (for message attachments) to 20 lists.
This number worked well for real data, but might not work well for other data.